### ISSUE TITLE
Implement levels.vctrs_vctr() if underlying type is character?

### ISSUE THREAD

**Issue by krlmlr** [r-lib/vctrs#1186]
Perhaps only with base type fallback, https://github.com/r-lib/vctrs/issues/1135.

With this, `rbind()` works well enough in this case.

``` r
library(vctrs)

foo <- new_vctr(letters, class = "foo")
data <- data.frame(foo)

class(data$foo)
#> [1] "foo"        "vctrs_vctr"

rbind(data, data)
#> Error: `levels.foo()` not supported.

levels.foo <- function(x) {
  NULL
}

class(rbind(data, data)$foo)
#> [1] "foo"        "vctrs_vctr"
```

<sup>Created on 2020-07-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by jennybc** [comment]
I independently ran across this and also pretty much convinced myself to implement `levels()` for my class, also motivated by making a simple `rbind()` work.

FWIW I **am** explicitly inheriting base type, i.e. the constructor uses `new_vctr(x, class = "drive_id", inherit_base_type = TRUE)`  and `rbind()` still doesn't work.

``` r
# I'm working on the drive_id class in a googledrive branch,
# but will soon merge this
devtools::load_all("~/rrr/googledrive")
#> ℹ Loading googledrive

x <- data.frame(a = as_id(month.name[1:3]))
x
#>          a
#> 1  January
#> 2 February
#> 3    March
x$a
#> <drive_id[3]>
#> [1] January  February March
class(x$a)
#> [1] "drive_id"   "vctrs_vctr" "character"

rbind(x, x)
#> Error: `levels.drive_id()` not supported.
```

<sup>Created on 2021-07-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

**Comment by jennybc** [comment]
But the more I look into it, I'm not so sure I *want* `rbind()` to "just work", because it uses `c()` and not `vec_c()`, so seems to circumvent many of the reasons for using vctrs to implement the `drive_id` class.

Maybe, all things considered, it's just as well that one needs to use `vctrs::vec_rbind()` or `dplyr::bind_rows()` in order to row bind data.frames hosting such a column.

**Comment by lionel-** [comment]
> But the more I look into it, I'm not so sure I want rbind() to "just work", because it uses c() and not vec_c(), so seems to circumvent many of the reasons for using vctrs to implement the drive_id class in the first place.

On the other hand, concatenating objects of the same class with `c()` should always work as expected. And by deriving from `vctrs_vctr` you automatically gain a `c()` method. I think the idea behind `vctrs_vctr` is to have the class work reasonably well with base, so it seems to make sense to implement `levels()` for vectors that explicitly inherit from `character`.

I'm just wondering what other consequences implementing `levels()` will have? I didn't know it was used for other things than factors.

**Comment by jennybc** [comment]
But, if we make `rbind()` "just work" by adding a placeholder method for `levels.foo`, then `rbind()` combines instances of `foo` in ways that violate what `vec_c()` would do.

Locally, I've added and exported `levels.drive_id` and here I show `rbind()` working to prove it:

``` r
# I'm working on the drive_id class in a googledrive branch,
# but will soon merge this
devtools::load_all("~/rrr/googledrive")
#> ℹ Loading googledrive

x <- data.frame(a = as_id(month.name[1:3]))
x
#>          a
#> 1  January
#> 2 February
#> 3    March

x$a
#> <drive_id[3]>
#> [1] January  February March
class(x$a)
#> [1] "drive_id"   "vctrs_vctr" "character"

# this works now, because I have a placeholder implementation of levels.drive_id
rbind(x, x)
#>          a
#> 1  January
#> 2 February
#> 3    March
#> 4  January
#> 5 February
#> 6    March
```

But here's what else "works" now. Notice how a combination of `drive_id` and `character` becomes `drive_id`, which is *not* what would happen with `vec_c()`. Am I just doing something wrong?

``` r
y <- data.frame(a = month.name[4:6])
y
#>       a
#> 1 April
#> 2   May
#> 3  June

y$a
#> [1] "April" "May"   "June"
class(y$a)
#> [1] "character"

z <- rbind(x, y)
z
#>          a
#> 1  January
#> 2 February
#> 3    March
#> 4    April
#> 5      May
#> 6     June
z$a
#> <drive_id[6]>
#> [1] January  February March    April    May      June
class(z$a)
#> [1] "drive_id"   "vctrs_vctr" "character"

xx <- vctrs::vec_c(as_id(month.name[1:3]),  month.name[4:6])
xx
#> [1] "January"  "February" "March"    "April"    "May"      "June"
class(xx)
#> [1] "character"
```

**Comment by DavisVaughan** [comment]
I don't believe that this is specific to character in any way. You can't use `rbind()` with a vctrs_vctr class with an underlying integer type either.

```r
library(vctrs)

data <- data.frame(foo = new_vctr(1:5, class = "foo"))

rbind(data, data)
#> Error: `levels.foo()` not supported.
```

Looking into this, here is where `levels()` is called once on each column of the data frame. It is a weird way to determine if something is factor-ish?? `xj` refers to a column in the data frame.
https://github.com/wch/r-source/blob/79298c499218846d14500255efd622b5021c10ec/src/library/base/R/dataframe.R#L1356-L1361

---

I think I would be ok with supplying a `levels()` method that returns `NULL`. I don't _think_ we'd want to rely on `levels.default` which returns `attr(x, "levels")`, as we'd want people to explicitly opt in to `levels()` support, but this gives us a decent default behavior that plays nicely with base R in the 99% case where there are no levels. I think we would continue to error with `levels<-.vctrs_vctr`.

I agree with Jenny that `rbind()` has some odd behavior that is different from `vec_rbind()` (see below), but I _think_ we'd probably prioritize compatibility with base R where possible, pointing people to use `vec_rbind()` if they complain about `rbind()` doing the "wrong thing". We do a similar thing with `c()` and `vec_c()`, so it seems consistent.

``` r
data <- data.frame(foo = new_vctr(1:5, class = "foo"))
data2 <- data.frame(foo = 1:5)

levels.vctrs_vctr <- function(x) {
  NULL
}

rbind(data, data)$foo
#> <foo[10]>
#>  [1] 1 2 3 4 5 1 2 3 4 5

rbind(data, data2)$foo
#> <foo[10]>
#>  [1] 1 2 3 4 5 1 2 3 4 5

# integer first, lost class
rbind(data2, data)$foo
#>  [1] 1 2 3 4 5 1 2 3 4 5
```

**Comment by jennybc** [comment]
> I agree with Jenny that `rbind()` has some odd behavior that is different from `vec_rbind()` (see below), but I think we'd probably prioritize compatibility with base R where possible

But in this case the base R results are wrong (or, rather, "can be and often are wrong"). So I'm not sure "compatibility" really describes what we'd be enabling.
